Improve `solc` compatibility

gnidan opened the pull request:
Re: #1 

- Use test from `truffle-contract` to test solc's compiled contract output
- ...

(WIP)

gnidan committed:
Migrate solc test from truffle-contract

gnidan commented:
@tcoulter is this the extent of the work for that ticket? the moved `truffle-contract` test now passes with minimal changes.

gnidan commented:
Notes from @tcoulter:
> @gnidan I’m thinking you should just rip out all this garbage and do what needs to happen. If you spend time on all this, we’ll likely have to rip it out again.
>
> > my thinking is that formalizing a schema warrants high-level discussion / design
>
> We already have an implicit schema used by 1000's of people. My gut says formalizing it is the way to get more input, since it’s not obvious what the schema entails currently.
> 
> In the future where a json schema exists, I’d likely want something like this:
> 
> 1. A function called `validate`, which takes in a json blob, throws that to jsonschema and calls its validate function.
> 
> 2. `n` number of `fromX` functions, like `fromSolidityStandardOutput`, which are helpers that output a validated JSON blob based on specific input.
> 
> Aside: I’m sorry if this conflicts with anything above. My thoughts are evolving as we talk about this more. (edited)

(Since then, we've decided to go with json-schema as part of this effort after all)

gnidan committed:
Refactor normalization / object generation

- Use separate methods for different sources (solc, abstraction, regular opts)

- Make `networks`, `address`, etc. (properties related to instance or
  truffle specifically) explicitly not a part of normalization process

- Move code around in `generateObject` to ensure truffle-option copying
  happens (now that it's gone from normalize)

- Switch to camelCase for fields in schema

- Add initial test for solc standard output

- Remove `x-` check for `normalizeInput`, as that's no longer part of that.

gnidan committed:
Refactor data normalization process

- Use property compatibility / data transformation mapping to identify
  where data should come from in a uniform manner.

gnidan commented on index.js:
```diff
@@ -21,132 +142,82 @@ var TruffleSchema = {
-    obj.contract_name = obj.contract_name || "Contract";
+    // if (options.overwrite == true) {
+    //   e = {};
+    // }
```
Oh I left this comment here - previously it was `existing_object = {}`, which wouldn't do anything, I would think except update the local variable to some other pointer. Is that overwrite behavior being relied on?

tcoulter replied:
I'm not entirely sure why the original line is here. I'd take it out (like you did) and obliterate the comment. There might be repercussions but hopefully we can find them. My guess is it's cruft from an old version of the artifactor.

gnidan committed:
Clean up properties descriptions a bit

gnidan committed:
Remove address from top-level schema def

gnidan committed:
Update property sourcing

- Drop notion of "additional sources" - define properties to list
  sources explicitly including their canonical name, or default to
  source only canonical name when `"sources":` is omitted

- Allow string values for sources in addition to functions. For
  string values, assume possible chain operation (e.g. "evm.bytecode.object")
  and split/chain accordingly

gnidan committed:
Clean-up property transformation magic

- Change `objectable` (whatever that means) to `objDirty`
- Move `properties` to the top and give it a docstring
- Move getter/chain combinators lower, also document

gnidan committed:
Remove meaningless comment block

gnidan committed:
Remove manual getter invocations from properties

gnidan commented:
Cleanup: merging this into `json-schema` to rework the corresponding PR #3 to include this (because more discussion is there)
